### DISCUSSION TITLE
Customize nitro build (and generate) command

### DISCUSSION THREAD

**Discussion by sebastianbayer** [nuxt/nuxt#19306] [Questions] [answered]
Hi

Im wondering how I would customize the nitro build/generate command the best way? As far as I read I would preferably do it in the `nuxt.config.ts` via the `nitro` property. But how exactely would I tell nitro when to execute a certain command? Would i use the nitro hooks for that?

I have the following problem: I download some assets while prerendering the routes and save them to the filesystem (public folder) during `generate`. Since the public folder is copied before prerendering (and downloading the assets) they are missing in the resulting output / dist folder. How can I hook into nitro so that it copies the public folder again after I prerendered all routes?

Thanks for your answers or any links to resources!

**Comment by Baroshem** [comment]
Hey,

I would recommend taking a look at the Nitro hooks that you can call from Nuxt app -> https://nuxt.com/docs/api/advanced/hooks#nitro-app-hooks-runtime-server-side

Or any from the Nuxt build hooks -> https://nuxt.com/docs/api/advanced/hooks#nuxt-hooks-build-time

**Reply by sebastianbayer** [reply]
Hi Baroshem – thanks for these resources! I now get how I would use hooks during generate but I don't yet understand exactely how I would adjust the generate command or nitro configs itself.


In my `nuxt.config.ts` I can make use of the `nitro:init` hook log the actual nitro instance with all its configs.
```
export default defineNuxtConfig({
  hooks: {
    'nitro:init': (nitro) => {console.log('nitro: ', nitro)}
  }
})

```

But how would I now change these configs in a way that I can repeat (or postpone) the execution `copyPublicAssets` (https://github.com/nuxt/nuxt/blob/main/packages/nuxt/src/core/nitro.ts#L239) function to be called at the end of prerendering all routes?

**Reply by Baroshem** [reply]
Maybe https://github.com/nuxt/nuxt/blob/3c006b33c8c6f9d73f456f0299d9ced6a07f4855/packages/schema/src/types/hooks.ts#LL197C9-L197C9

**Reply by sebastianbayer** [reply]
Alright! I will try using the `nitro:config` to interact with directly with nitro and extend configs to call functions at certain points (hooks) during the `nuxt generate` command…

Did you – by any chance – find any examples for this or could you provide a simple one?

**Comment by sebastianbayer** [comment, marked as answer]
I ended up creating a local nuxt module that hooks into the nuxt close hook and then calls the `copyPublicAssets` function from nitro to copy the public folder again after prerendering the routes with the crawler (and downloading the assets mentioned in the first question).

The module (`copyPublicFolder.ts`) looks like this:

```
import { defineNuxtModule } from '@nuxt/kit'
import { createNitro, copyPublicAssets } from 'nitropack'

export default defineNuxtModule({
  meta: {
    name: '@nuxtjs/test',
    configKey: 'sample',
    compatibility: {
      nuxt: '^3.0.0'
    }
  },
  defaults: {},
  hooks: {},
  setup (options, nuxt) {
    nuxt.hook('close', async () => {
        const nitro = await createNitro()
        await copyPublicAssets(nitro)
    })
  }
})
```

in `nuxt.config.ts` i use the module like this:

```
export default defineNuxtConfig({
  modules: ['./modules/copyPublicFolder'],
})
```
